addons/web_editor/static/src/components/media_dialog/image_selector.js

```diff
@@ -0,0 +1,261 @@
+        }
+        const { media } = await this.fetchLibraryMedia(this.state.libraryMedia.length);
+        this.state.libraryMedia.push(...media);
+    }
+
+    async search(...args) {
+        await super.search(...args);
+        if (!this.props.useMediaLibrary) {
+            return;
+        }
+        if (!this.state.needle) {
+            this.state.searchService = 'all';
+        }
+        const { media, results } = await this.fetchLibraryMedia(0);
+        this.state.libraryMedia = media;
+        this.state.libraryResults = results;
+    }
+
+    async onClickAttachment(attachment) {
+        this.selectAttachment(attachment);
+        if (!this.props.multiSelect) {
+            await this.props.save();
+        }
+    }
+
+    async onClickMedia(media) {
+        this.props.selectMedia({ ...media, mediaType: 'libraryMedia' });
+        if (!this.props.multiSelect) {
+            await this.props.save();
+        }
+    }
+
+    /**
+     * Utility method used by the MediaDialog component.
+     */
+    static async createElements(selectedMedia, { orm, rpc }) {
+        // Create all media-library attachments.
+        const toSave = Object.fromEntries(selectedMedia.filter(media => media.mediaType === 'libraryMedia').map(media => [
+            media.id, {
+                query: media.query || '',
+                is_dynamic_svg: !!media.isDynamicSVG,
+                dynamic_colors: media.dynamicColors,
+            }
+        ]));
+        let savedMedia = [];
+        if (Object.keys(toSave).length !== 0) {
+            savedMedia = await rpc('/web_editor/save_library_media', { media: toSave });
+        }
+        const selected = selectedMedia.filter(media => media.mediaType === 'attachment').concat(savedMedia).map(attachment => {
+            // Color-customize dynamic SVGs with the theme colors
+            if (attachment.image_src && attachment.image_src.startsWith('/web_editor/shape/')) {
+                const colorCustomizedURL = new URL(attachment.image_src, window.location.origin);
+                colorCustomizedURL.searchParams.forEach((value, key) => {
+                    const match = key.match(/^c([1-5])$/);
+                    if (match) {
+                        colorCustomizedURL.searchParams.set(key, getCSSVariableValue(`o-color-${match[1]}`));
+                    }
+                });
+                attachment.image_src = colorCustomizedURL.pathname + colorCustomizedURL.search;
+            }
+            return attachment;
+        });
+        return Promise.all(selected.map(async (attachment) => {
+            const imageEl = document.createElement('img');
+            let src = attachment.image_src;
+            if (!attachment.public) {
+                const [accessToken] = await orm.call(
+                    'ir.attachment',
+                    'generate_access_token',
+                    [attachment.id],
+                );
+                src += `?access_token=${accessToken}`;
```

Comment:
the remark about url-encoding is true everywhere URLs are crafted with string formatting instead of using `URLSearchParams` or its various alternatives. In most cases it may be safe... until one day it isn't. It's just a best practice to follow.

Reply:
As discussed, we'll track down the occurrences of such mistakes in stable (that one here already existed before this PR) and forward-port correctly up to master. I am adding an entry in our bug-fix pad.